src/EventStore.Core/Services/Transport/Grpc/Redaction.SwitchChunk.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Threading.Tasks;
+using EventStore.Client.Redaction;
+using EventStore.Core.Data.Redaction;
+using EventStore.Core.Messages;
+using EventStore.Core.Messaging;
+using EventStore.Plugins.Authorization;
+using Grpc.Core;
+
+namespace EventStore.Core.Services.Transport.Grpc {
+	internal partial class Redaction {
+		private static readonly Operation SwitchChunkOperation = new(EventStore.Plugins.Authorization.Operations.Node.Redaction.SwitchChunk);
+
+		public override async Task SwitchChunk(
+			IAsyncStreamReader<SwitchChunkReq> requestStream,
+			IServerStreamWriter<SwitchChunkResp> responseStream,
+			ServerCallContext context) {
+
+			var user = context.GetHttpContext().User;
+			if (!await _authorizationProvider.CheckAccessAsync(user, SwitchChunkOperation, context.CancellationToken).ConfigureAwait(false))
+				throw RpcExceptions.AccessDenied();
+
+			await SwitchChunksLock().ConfigureAwait(false);
+			try {
+				await SwitchChunks(context, requestStream, responseStream).ConfigureAwait(false);
+			} finally {
+				await SwitchChunksUnlock().ConfigureAwait(false);
+			}
+		}
+
+		private async Task SwitchChunksLock() {
+			var lockTcs = new TaskCompletionSource<RedactionMessage.SwitchChunkLockCompleted>();
+			_bus.Publish(new RedactionMessage.SwitchChunkLock(
+				new CallbackEnvelope(msg => lockTcs.SetResult(msg as RedactionMessage.SwitchChunkLockCompleted))
+			));
+
+			var completionMsg = await lockTcs.Task.ConfigureAwait(false);
+
+			// it's important not to throw any exceptions beyond this point in this method
+			// (except to indicate that acquiring the lock failed), otherwise we can end up
+			// in a situation where the lock remains permanently acquired.
+
+			if (completionMsg is null) // should never happen
+				return;
```

Comment:
probably throw an exception

Reply:
:+1:

Reply:
fixed in 1f7be5d0cfc215e2c3c0cb0a511e01f31d5071c8 but eventually overridden by TcsEnvelope changes (that throws if the resulting message type is not correct)